crates/cow-amm/src/lib.rs

```diff
@@ -0,0 +1,26 @@
+mod event_updater;
+mod implementations;
+mod registry;
+
+use {ethcontract::Address, std::sync::Arc};
+pub use {
+    implementations::safe_based::event_handler::Contract as CowAmmSafeBasedContract,
+    registry::Registry,
+};
+
+pub trait CowAmm: Send + Sync {
+    /// Address of the CoW AMM.
+    /// Can be used by the autopilot to build the list of accepted cow amms.
+    fn address(&self) -> &Address;
+
+    /// Returns the list of tokens traded by this pool.
+    /// Can be used by the autopilot to build the list of native token prices to
+    /// query.
+    fn traded_tokens(&self) -> &[Address];
+}
+
+#[async_trait::async_trait]
+pub trait ContractHandler: Sync + Send {
+    /// Apply the event to the given CoW AMM registry
+    async fn apply_event(&self) -> anyhow::Result<Option<Arc<dyn CowAmm>>>;
+}
```

Comment:
This trait feels a bit weird. I experimented with changing the trait like this as this better expresses what concept we really want to express (some contract that behaves like a factory where we can determine whether an event deployed a new cow amm):
```rust
/// A contract which serves as a factory for new CoW AMMs (i.e. deploys them).
pub trait Factory: EventRetrieving + Send + Sync + 'static {
    /// Returns whether the given contract deployed a new cow amm.
    fn deployment(event: &Self::Event) -> Option<Arc<dyn CowAmm>>;
}
```

IMO this is not a clear win so please take a look and let me know what you think.
[trait_experiment.txt](https://github.com/user-attachments/files/15920762/trait_experiment.txt)

Reply:
@MartinquaXD I don't see what is the benefit of your proposal, and what feels weird about the trait I wrote. The trait can be documented better and can be restrained more, but besides that I don't see any issue with it.

With your solution we are adding a very important constraint:

```
pub cow_amm_indexer: cow_amm::Registry<cow_amm::CowAmmSafeBasedContract>,
```

so basically all the abstraction I did is for nothing since the registry is built for an specific event. As I did it, you build the registry and it contains a registry all the events which fulfill the trait.